Global Loot Funcs on 1.15.x

Draco18s opened the pull request:
See #6267 for complete details.

Data driven system for arbitrary loot modification as a replacement for HarvestDropsEvent, addressing #5871.

Draco18s committed:
Global Loot Funcs on 1.15.x

metelares commented:
change e-mail

Draco18s commented:
Hey @LexManos just saw the LTS post on the forums, your request for me to target 1.15 in order to get this merged makes a lot more sense now. 👍 Just let me know if anything needs to be changed and I'll get it done asap.

@metelares huh? Your comment makes no sense.

LexManos commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1210,4 +1210,12 @@ public static synchronized void updateBurns()
 
+    public static List<ItemStack> modifyLoot(List<ItemStack> list, LootContext context) {
+        LootModifierManager man = LootModifierManager.getInstance();
+        for(IGlobalLootModifier mod : man.getAllLootMods()) {
```
This doesn't seem to address re-processing new list entries added by modifiers.
Such as re-running a smelting modifier after modifier that converts gravel to sand.
Is that something we want to care about at all?

Draco18s replied:
I don't know if we want to handle it or not. I'm sure someone will eventually have a use-case, but it was not in the list of things people cared about when I was writing the system. Such processing would have to cleanly handle probabalistic alterations (e.g. if gravel is converted to sand 10% of the time (first pass: fail) and the list gets re-processed, it should not attempt to recheck that element). I'm worried that this becomes non-trivial quickly.

LexManos replied:
That is the entire issue of why I didn't work on this yet.
But, because things are being designed, and this is most definitely going to be a breaking change {processing entries, instead of the full list} it needs to be addressed.
Perhaps the processor should be `List<ItemStack> process(ItemStack);`

Draco18s replied:
I have a specific use-case that requires the full list, not just individual entries. One of the test mods (`WheatSeedsConverterModifier`) shows one such an example.

LexManos commented on src/main/java/net/minecraftforge/common/loot/IGlobalLootModifier.java:
```diff
@@ -0,0 +1,10 @@
+import net.minecraft.world.storage.loot.LootContext;
+
+public interface IGlobalLootModifier {
+    List<ItemStack> apply(List<ItemStack> generatedLoot, LootContext context);
```
This seems to be the core of the system, yet ZERO docs?

Draco18s replied:
I'll address this. I didn't even want to open a PR originally because I was bashing around blindly, but I got zero feedback.

LexManos commented on src/main/java/net/minecraftforge/common/loot/IGlobalLootModifierSerializer.java:
```diff
@@ -0,0 +1,17 @@
+import net.minecraft.world.storage.loot.conditions.ILootCondition;
+import net.minecraftforge.registries.IForgeRegistryEntry;
+
+public interface IGlobalLootModifierSerializer<T extends IGlobalLootModifier> extends IForgeRegistryEntry<IGlobalLootModifierSerializer<?>> {
```
Again, docs.

LexManos commented on src/main/java/net/minecraftforge/common/loot/IGlobalLootModifierSerializer.java:
```diff
@@ -0,0 +1,17 @@
+    public abstract T read(ResourceLocation location, JsonObject object, ILootCondition[] ailootcondition);
+
+    @Override
+    public default Class<IGlobalLootModifierSerializer<?>> getRegistryType() {
```
This needs a valid value.

Draco18s replied:
Good point. I looked at it a day or so ago (I saw that it was returning null and went "this probably should not do that") but I have been unable to determine what it *should* return. The existing implementations don't help (there being only two: one is `(Class<V>)token.getRawType();` and the other is `Serializer.<IRecipeSerializer<?>>castClass(IRecipeSerializer.class);`, attempting to mimic either one just generates compile-time errors).

LexManos replied:
The one to mimic, due to generics, would be IRecipeSerializer. But you may have to do some finessing with generics to get it correct. Or, you could just make IGlobalLootModifierSerializer non-generic.

Draco18s replied:
Yeah, I was trying to play with the IRecipeSerializer one, but it kept fighting me. IRecipeSerializer contains within it a `Serializer` definition, which I don't have.

Draco18s replied:
Figured this one out. Involved moving the functionality down to the base class instead of in the interface `default`, but that's fine. There had been a part of the implementation in IRecipeSerializer that I overlooked.

LexManos commented on src/main/java/net/minecraftforge/common/loot/LootModifierManager.java:
```diff
@@ -0,0 +1,80 @@
+        instance = this;
+    }
+
+    public static LootModifierManager getInstance() {
```
This system is completely messed up, you're creating a new instance every time a MinecraftServer is created, such as whenever a single player world is loaded.
Better to stop using a static context and instead give some other way of getting the manager.

Draco18s replied:
The only other option is to request it from the MinecraftServer instance. Which I didn't do as it would both make it annoying to get at or make the patch to MinecraftServer larger. I'm open to alternatives.

LexManos replied:
What would it take to get the server instance to where it needs to be?
And the 'larger' patch would just be a 2 line accessor.

Draco18s replied:
MinecraftServer does not have an accessor except via a `World` object (and a `ServerWorld` at that). I'm fine going that route, but I'm in the dark about what's "best practice" here.

LexManos replied:
If world accessible in context? Then ya, go through it. Going through several getters isn't that big of a deal.

LexManos commented on src/main/java/net/minecraftforge/common/loot/LootModifierManager.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Override
+    protected void apply(Map<ResourceLocation, JsonObject> splashList, IResourceManager resourceManagerIn, IProfiler profilerIn) {
```
This is a custom class, which means we don't have to use the same parameter names. So... use real names here... splashList is DEFINITLY not a list of splashes

LexManos commented on src/test/java/net/minecraftforge/debug/gameplay/loot/GlobalLootModifiersTest.java:
```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * The smelting enchantment causes this modifier to be invoked, via the smelting loot_modifier json
+     * @author Draco18s
```
no

Draco18s replied:
No *what?* "Don't put javadoc" "don't put @author"? What am I doing wrong?

LexManos replied:
I did ti specifically on the author line, so it's about the @author, we don't do that in Forge.
Javadocs are good, Ego stroking is bad. {Not saying that's what you're doing but you get enough credit in the git log it doesn't need to be in the source}

Draco18s replied:
The comment widget includes four lines of code and doesn't indicate exactly which line you pinged. Removing @author, got it. Thanks. I didn't put it to stoke my ego, it was added automatically when I inserted the javadoc.

LexManos replied:
Just for future reference, the widget will always show on the line that it was commented on, the 3 lines above it are for context.

Draco18s replied:
Fair enough. Just with the Github UI was a little better, I guess. Anyway, removing those lines is easy enough.

LexManos commented on src/test/java/net/minecraftforge/debug/gameplay/loot/GlobalLootModifiersTest.java:
```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * When harvesting blocks with bamboo, this modifier is invoked, via the silk_touch_bamboo loot_modifier json
+     * @author Draco18s
```
No

LexManos commented on src/test/java/net/minecraftforge/debug/gameplay/loot/GlobalLootModifiersTest.java:
```diff
@@ -0,0 +1,193 @@
+            ItemStack ctxTool = context.get(LootParameters.TOOL);
+            ItemStack fakeTool = ctxTool.copy();
+            fakeTool.addEnchantment(Enchantments.SILK_TOUCH, 1);
+            if(ItemStack.areItemsEqual(ctxTool, fakeTool) && EnchantmentHelper.getEnchantments(ctxTool).containsKey(Enchantments.SILK_TOUCH)) return generatedLoot;
```
What is the point of this?

Draco18s replied:
This goes back to comments @williewillus made. Boils down to "you can't just get the block's item form."
https://github.com/MinecraftForge/MinecraftForge/pull/5871#issuecomment-542738660

If I've implemented *a test mod* incorrectly, I'm not sure what to say. Its a stand-in for something a modder would do, its not shipped code.

LexManos replied:
Test mods also are used as examples, and unfortunately copy/pasted by idiot modders.
Perhaps this particular modifier would be better served as a pre-processor, something that can always add the enchantment for silk touch to the context? Instead of trying to figure out how to reproduce something after the fact.

Draco18s replied:
LootContext objects are `private final`, so there's no way to "add" to it. Yes, it could be modified with reflection, but I don't think that's the appropriate action to take.
(But I'm fine being wrong)

LexManos commented on src/test/java/net/minecraftforge/debug/gameplay/loot/GlobalLootModifiersTest.java:
```diff
@@ -0,0 +1,193 @@
+                    .withNullableParameter(LootParameters.BLOCK_ENTITY, context.get(LootParameters.BLOCK_ENTITY));
+            if(context.has(LootParameters.THIS_ENTITY)) {
+                builder.withParameter(LootParameters.THIS_ENTITY, context.get(LootParameters.THIS_ENTITY));
+            }
```
This seems super bad, why are you rebuilding the entire context? How can you possibly know the context and what parts need copying?

Draco18s replied:
Again, test mod code, not shippable.

Draco18s replied:
And if there's a better way, I'd be happy to do it. This was what came up in the prior conversation. LootContext objects are `private final` and the context *has* to be cloned in order for the functionality to work, as the original context should not be modified (as it will get passed to future LootModifier handlers). If that means we need to patch in a Builder.FromContext of some kind, then fine, I'll do that.

LexManos reviewed (changes requested):
This isn't bad, my main concerns are the lack of documentation on the root internfaces.
I also don't see where the serialization comes into play.
And lastly, one of the main reasons I've been procrastinating doing this. Do we just not care about reentrancy at all?

Draco18s commented:
Other comments re @LexManos 

> I also don't see where the serialization comes into play.

It probably doesn't. Any code that still exists as a vestige of my hacking about semi-blindly. I'm not even sure what parts that might be, the interface only contains a `read` method, no `write`. If I missed something, point it out.

> And lastly, one of the main reasons I've been procrastinating doing this. Do we just not care about reentrancy at all?

I don't know, do we? This is one of the reasons I tried to request feedback from you two months ago when I started working on this. The only comment in this regard was about what happens if two mods want to modify the same drop differently, what order takes precedence, which is why the list is sorted by registry name.

If we *do* care, we also have to address `A->B` + `A->C` precedence handling, not just `B->C` + `A->B`. The alternative is to *not* care, which has precedent (there can not be an "absolutely last, no matter what" event handler).

I wrote this PR because no one else was, I don't know what I'm doing, I'm just trying to get *something* done so that I (and many others) can move forwards.

Draco18s committed:
Updates based on Lex's comments

Removed @author
Added a LootContext.Builder constructor that can clone a context object
Added javadoc
Made LootModifierManager not a singleton

Draco18s committed:
Missing json files

Draco18s committed:
Refactor to read layered config file

Reads all.json to determine what loot modifiers to load. Functions like tags.

LexManos commented:
We had a long discussion and brainstorming session on discord, he knows what he needs to do to move forward. Marked this as WIP until he's done.

Draco18s committed:
Further refactor

Updated EntityLiving's call
Updated LootTable to mark previous entityliving entry as deprecated
Renamed all.json to global_loot_modifiers.json
Relocated test mod data resources to generated_test
Fixed setRegistryName implementation
Tabs to spaces
Refactored json loading system making it simpler and less round about.

Draco18s committed:
Docs clean

Cleaned up javadoc wording in several places
Made setRegistryName final

Draco18s committed:
Wordsmithing

Adjusted javadoc, added @nonnull annotations

Draco18s committed:
Merge branch '1.15.x' into Global-Loot-Funcs-1.15

# Conflicts:
#	patches/minecraft/net/minecraft/entity/LivingEntity.java.patch
#	patches/minecraft/net/minecraft/server/MinecraftServer.java.patch
#	src/test/resources/META-INF/mods.toml

Draco18s commented:
All requested changes have been made. I have requested a re-review. What is holding this up?

Draco18s committed:
Merge branch '1.15.x' into Global-Loot-Funcs-1.15

gigaherz commented on src/main/java/net/minecraftforge/common/loot/LootModifier.java:
```diff
@@ -0,0 +1,105 @@
+     * <code>read</code> method to deserialize from json.
+     * @param <T> the final Type
+     */
+    public abstract static class Serializer<T extends LootModifier> implements IGlobalLootModifierSerializer<T> {
```
Shouldn't this be extending `ForgeRegistryEntry<IGlobalLootModifierSerializer<?>>` instead of implementing `IForgeRegsitryEntry<T>` by hand?

Actually, I wonder if it wouldn't be best to make `IGlobalLootModifierSerializer` an abstract class instead, then subclasses wouldn't need to worry about IForgeRegsitryEntry

Draco18s replied:
Hmm, cursory check shows that that is doable (it ran and didn't crash, can't check more than that at this instant).

Draco18s replied:
Yep, it works fine that way (had just wanted to verify that things were still being loaded and the test mod working). I personally don't care if its an interface or an abstract class, but it does seem cleaner/easier on the modder as the abstract class.

gigaherz replied:
The main reason to use an interface over an abstract class, is that you don't force users into a specific implementation, since they can implement that interface anywhere they like.
However I'd argue that in this case, there isn't much point of keeping it an interface? People shouldn't be making any random object their loot modifier serializer.

Draco18s replied:
Yep, agreed. There was nothing in the abstract implementation that a modder needed to touch, so I pushed it upwards and made it final.

gigaherz reviewed (commented):
I had completely disregarded this PR thinking it was just some other "let's add new loot events!" proposal, and I'm pleasantly surprised of what I see.

I do have one concern, but don't think of it as a change request, rather only as a discussion topic.

Draco18s committed:
Changed from Interface to Abstract Class

GlobalLootModifierSerializer is now an abstract class and the abstract implementation in LootModifier was no longer needed.

Draco18s committed:
Stray tab

LexManos reviewed (approved):
Few code style issues, and still lacking documentation in places. But it'll work.

Draco18s commented:
I'll backport this to 1.14 this weekend. I'll see if I can do something about docs and style, but I feel like I've been over them a bunch and am not sure what to add.

dhharris commented:
@Draco18s , I'm looking for a HarvestDropsEvent replacement so I can keep track of loot mined. Does this feature support that?
I'm looking to track loot _without_ modifying it. For example, you mine a redstone ore, and during the event it updates a class with the BlockPos and number of redstone dropped.

Draco18s commented:
You'd still have to register it *as* a modifier, but you don't have to actually *modify* anything (and you'd still get the list of harvested loot and full context).
